prow/crier/controller_test.go

```diff
@@ -0,0 +1,206 @@
+	tr *testResults
+}
+
+func (f triggerReporter) Report(pj *apiv1.ProwJob) ([]*apiv1.ProwJob, error) {
+	f.tr.mux.Lock()
+	if pj.Status.State == apiv1.TriggeredState {
+		f.tr.reporterPasses = true
+	} else {
+		f.tr.reporterFails = true
+	}
+	f.tr.mux.Unlock()
+	return nil, nil
+}
+
+func (f triggerReporter) GetName() string {
+	return "triggerReporter"
+}
+
+func (f triggerReporter) ShouldReport(pj *apiv1.ProwJob) bool {
+	return pj.Status.State == apiv1.TriggeredState
+}
+
+//Fake Informer
+//Sets: The Prow Job Test Cases
+//Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	tr *testResults
+}
+
+func (f fakeInformer) Get(name string) (*apiv1.ProwJob, error) {
+	var pj *apiv1.ProwJob
+	switch name {
+	case "TestJob1":
+		pj = &apiv1.ProwJob{
+			Spec: apiv1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: apiv1.ProwJobStatus{
+				State: apiv1.TriggeredState,
+			},
+		}
+	case "TestJob2":
+		pj = &apiv1.ProwJob{
+			Spec: apiv1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: apiv1.ProwJobStatus{
+				State: apiv1.FailureState,
+			},
+		}
+	}
+
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) listerv1.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() listerv1.ProwJobLister {
+	return f
+}
+
+func (f fakeInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	f.tr.mux.Lock()
+	f.tr.isRegistered = true
+	f.tr.mux.Unlock()
+}
+
+func (f fakeInformer) Run(stopCh <-chan struct{}) {
+	f.tr.mux.Lock()
+	f.tr.isRun = true
+	f.tr.mux.Unlock()
+}
+
+func (f fakeInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	f.tr.mux.Lock()
+	f.tr.isRegistered = true
+	f.tr.mux.Unlock()
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) GetStore() cache.Store                                           { return nil }
+func (fakeInformer) GetController() cache.Controller                                 { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                 { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                       { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                       { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*apiv1.ProwJob, err error) { return nil, nil }
+
+func TestController_Run(t *testing.T) {
+
+	//Initialization
+	cs := fake.NewSimpleClientset()
+
+	q := kube.RateLimiter("CrierTest")
+	q.Add("TestJob1")
+	q.Add("TestJob2")
+
+	var results testResults
+
+	inf := fakeInformer{}
+	inf.tr = &results
+
+	rp := triggerReporter{}
+	rp.tr = &results
+
+	nmwrk := 2
+
+	wg := &sync.WaitGroup{}
+
+	c := NewController(cs, q, inf, rp, nmwrk, wg)
+
+	//Testing
+	stopCh := make(chan struct{})
+	go c.Run(stopCh)
```

Comment:
hummm - if you already have the queue pre-populated, I'm thinking of maybe you can just call `c.runWorker()` here directly until it returns, without a separate go-routine and the polling logic, wdyt

Reply:
I could do that, and it would be simpler, but I prefer this, as it's:
1) testing a public method (Run) instead of a private method (runWorker)**
2) testing the goroutine creation code within controller.go

**Checking the queue's length to determine when to quit is an encapsulation violation, but that's because these controllers aren't built to signal or halt when their queue is empty; they run "forever". I recognize we have to violate encapsulation either way

Reply:
Actually, should there be a series of unit tests for runWorker _and_ an integration test for Run?

Reply:
that also sgtm

Reply:
Does decoupling the tests (most recent commit) make the tests any easier to understand?

Reply:
Now that the test has been switched over to table-driven development, it's obvious to me that decoupling improves the test